### PR TITLE
Flesh out pubsub subscriptions

### PR DESCRIPTION
**Pull request by tseaver**
Uses #742 as a base.

### CODE REVIEW

**Comment by dhermes** [comment]
@tseaver I'm going to hold off on this until #742 is through. Is that OK or do you prefer otherwise?

**Comment by tseaver** [comment]
Holding off here is fine.

**Comment by tseaver** [comment]
@dhermes rebased after merge of #742.

File: gcloud/pubsub/api.py [hunk 1]
```diff
@@ -52,3 +52,49 @@ def list_topics(page_size=None, page_token=None,
 
     path = '/projects/%s/topics' % project
     return connection.api_request(method='GET', path=path, query_params=params)
+
+
+def list_subscriptions(page_size=None, page_token=None, topic_name=None,
+                       project=None, connection=None):
+    """List subscriptions for a given project.
+
```
Comment (dhermes) [inline gcloud/pubsub/api.py]: Can you include the resource->method it maps to in
https://www.googleapis.com/discovery/v1/apis/pubsub/v1beta2/rest

It seems they are

```
pubsub.projects.subscriptions.list
pubsub.projects.topics.subscriptions.list
```

File: gcloud/pubsub/subscription.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    :type name: string
+    :param name: the name of the subscription
+
+    :type topic: :class:`gcloud.pubsub.topic.Topic`
+    :param topic: the topic to which the subscription belongs..
+
+    :type ack_deadline: int
+    :param ack_deadline: the deadline (in seconds) by which messages pulled
+                         from the back-end must be ACKed.
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: ACKed -> acknowledged. (I suppose they could just click the link to learn more).

File: gcloud/pubsub/subscription.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        self.ack_deadline = data.get('ackDeadline')
+        push_config = data.get('pushConfig', {})
+        self.push_endpoint = push_config.get('pushEndpoint')
+
+    def modify_push_configuration(self, push_endpoint):
+        """API call:  update the push endpoint for the subscription.
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/subscriptions/modifyPushConfig
+        """
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: No args or returns.

File: gcloud/pubsub/subscription.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            return False
+        else:
+            return True
+
+    def reload(self):
+        """API call:  test existence of the subsription via a GET request
+
+        See
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/subscriptions/get
+        """
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Needs args and returns

**Comment by dhermes** [comment]
LGTM though you should get sign off from @tmatsuo to make sure you are making the correct API calls

File: gcloud/pubsub/api.py [hunk 5]
```diff
@@ -52,3 +52,49 @@ def list_topics(page_size=None, page_token=None,
...
+    params = {}
+
+    if page_size is not None:
+        params['pageSize'] = page_size
+
+    if page_token is not None:
+        params['pageToken'] = page_token
+
+    if topic_name is None:
+        path = '/projects/%s/subscriptions' % project
```
Comment (tmatsuo) [inline gcloud/pubsub/api.py]: What happens if the project is None?

File: gcloud/pubsub/subscription.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            conn.api_request(method='GET',
+                             path=self.path,
+                             query_params={'fields': 'name'})
+        except NotFound:
+            return False
+        else:
+            return True
+
+    def reload(self):
+        """API call:  test existence of the subsription via a GET request
```
Comment (tmatsuo) [inline gcloud/pubsub/subscription.py]: wrong docstring

File: gcloud/pubsub/subscription.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        :param ack_ids: ack IDs of messages being acknowledged
+        """
+        data = {'ackIds': ack_ids}
+        conn = self.topic.connection
+        conn.api_request(method='POST',
+                         path='%s:acknowledge' % self.path,
+                         data=data)
+
+    def modify_ack_deadline(self, ack_id, ack_deadline):
+        """API call:  acknowledge retrieved messages for the subscription.
```
Comment (tmatsuo) [inline gcloud/pubsub/subscription.py]: wrong docstring

**Comment by tmatsuo** [comment]
Left few comments, but LGTM in terms of the correctness of API paths and methods.

File: gcloud/pubsub/api.py (hunk 1 above)
Comment (tseaver) [inline gcloud/pubsub/api.py]: db75350

File: gcloud/pubsub/subscription.py (hunk 2 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: bdcc60b

File: gcloud/pubsub/subscription.py (hunk 3 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: e1b581dddefb3130d023d29da9809a69c46edea9

File: gcloud/pubsub/api.py (hunk 5 above)
Comment (tseaver) [inline gcloud/pubsub/api.py]: `project` will never be None once we are done folding in the "implicit environment" stuff  we will update in a subsequent PR to infer the "default" project from the environment, if it is not passed explicitly to the topic:  at that point, if the inference fails, we will raise an exception from `Topic.__init__`.

File: gcloud/pubsub/subscription.py (hunk 6 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: 6a9eb19

File: gcloud/pubsub/subscription.py (hunk 7 above)
Comment (tseaver) [inline gcloud/pubsub/subscription.py]: ea1625c

File: gcloud/pubsub/api.py (hunk 1 above)
Comment (dhermes) [inline gcloud/pubsub/api.py]: Those links are fine but I literally meant using the resource names I included above. That's the way  to refer to these API methods. It's fine as is though.

File: gcloud/pubsub/subscription.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        :param ack_ids: ack IDs of messages being acknowledged
+        """
+        data = {'ackIds': ack_ids}
+        conn = self.topic.connection
+        conn.api_request(method='POST',
+                         path='%s:acknowledge' % self.path,
+                         data=data)
+
+    def modify_ack_deadline(self, ack_id, ack_deadline):
+        """API call:  update acknowledgement deadline for a retrieved messages.
```
Comment (tmatsuo) [inline gcloud/pubsub/subscription.py]: "a retrieved message."

File: gcloud/pubsub/subscription.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        :param ack_deadline: new deadline for the message, in seconds
+        """
+        data = {'ackId': ack_id, 'ackDeadlineSeconds': ack_deadline}
+        conn = self.topic.connection
+        conn.api_request(method='POST',
+                         path='%s:modifyAckDeadline' % self.path,
+                         data=data)
+
+    def delete(self):
+        """API call:  delete the subscription via a DELETE request
```
Comment (tmatsuo) [inline gcloud/pubsub/subscription.py]: period at the end?

**Comment by tmatsuo** [comment]
still LGTM

**Pull request merged**